File: XCTestBootstrap/Utility/FBOToolOperation.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBOToolOperation.h"

#import "XCTestBootstrapError.h"

@implementation FBOToolOperation

+ (FBFuture<NSArray<NSString*>*>*)listSanitiserDylibsRequiredByBundle:(NSString*)testBundlePath onQueue:(nonnull dispatch_queue_t)queue {
  NSBundle *bundle = [NSBundle bundleWithPath:testBundlePath];
  if (!bundle) {
    NSString *message = [NSString stringWithFormat:@"Bundle '%@' does not identify an accessible bundle directory.", testBundlePath];
    return [FBFuture futureWithError:[[XCTestBootstrapError describe:message] build]];
  }
  if (![bundle executablePath]) {
    NSString *message = [NSString stringWithFormat:@"The bundle at %@ does not contain an executable.", testBundlePath];
    return [FBFuture futureWithError:[[XCTestBootstrapError describe:message] build]];
  }
  
  return [[[[[[FBProcessBuilder
    withLaunchPath:@"/usr/bin/otool"]
    withArguments:@[@"-L", [bundle executablePath]]]
    withStdOutInMemoryAsString]
    withStdErrInMemoryAsString]
    runUntilCompletionWithAcceptableExitCodes:[NSSet setWithObject:@0]]
    onQueue:queue map:^ NSArray<NSString * > *(FBProcess<id, NSString *, NSString *> *task) {
      return [FBOToolOperation extractSanitiserDylibsfromOtoolOutput:task.stdOut];
    }];
}

+ (NSArray<NSString *> *)extractSanitiserDylibsfromOtoolOutput:(NSString *)otool_output {
  NSRegularExpression *regex = [NSRegularExpression regularExpressionWithPattern:@"@rpath/(libclang_rt\\..*san_.*_dynamic.dylib)"
                                                                         options:NSRegularExpressionCaseInsensitive
                                                                           error:nil];
  NSMutableArray *libs = NSMutableArray.array;
  
  [regex enumerateMatchesInString:otool_output
                          options:0
                            range:NSMakeRange(0, otool_output.length)
                       usingBlock:^(NSTextCheckingResult * _Nullable result, NSMatchingFlags flags, BOOL * _Nonnull stop) {
    NSRange range = [result rangeAtIndex:1];
    NSString *libName = [otool_output substringWithRange:range];
    [libs addObject:libName];
  }];
  return [NSArray arrayWithArray:libs];
}

@end

File: XCTestBootstrap/Utility/FBOToolOperation.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import FBControlCore
import Foundation

public enum FBOToolError: Error {
  case bundleInaccessible(path: String)
  case bundleMissingExecutable(path: String)
}

extension FBOToolError: LocalizedError {
  public var errorDescription: String? {
    switch self {
    case let .bundleInaccessible(path):
      return "Bundle '\(path)' does not identify an accessible bundle directory."
    case let .bundleMissingExecutable(path):
      return "The bundle at \(path) does not contain an executable."
    }
  }
}

public final class FBOToolOperation {

  public static func listSanitiserDylibsRequired(byBundle testBundlePath: String) async throws -> [String] {
    guard let bundle = Bundle(path: testBundlePath) else {
      throw FBOToolError.bundleInaccessible(path: testBundlePath)
    }
    guard let executablePath = bundle.executablePath else {
      throw FBOToolError.bundleMissingExecutable(path: testBundlePath)
    }

    let result = try await Subprocess(executable: "/usr/bin/otool", arguments: ["-L", executablePath])
      .run()
    return extractSanitiserDylibs(fromOtoolOutput: result.standardOutput)
  }

  private static func extractSanitiserDylibs(fromOtoolOutput otoolOutput: String) -> [String] {
    guard let regex = try? NSRegularExpression(pattern: "@rpath/(libclang_rt\\..*san_.*_dynamic.dylib)", options: .caseInsensitive) else {
      return []
    }
    var libs: [String] = []
    let nsString = otoolOutput as NSString
    regex.enumerateMatches(in: otoolOutput, options: [], range: NSRange(location: 0, length: nsString.length)) { result, _, _ in
      guard let result else { return }
      let range = result.range(at: 1)
      libs.append(nsString.substring(with: range))
    }
    return libs
  }
}
